README
NAME
      Crypt::SSLeay - OpenSSL glue that provides LWP https support

SYNOPSIS
      lwp-request https://www.example.com

      use LWP::UserAgent;
      my $ua = new LWP::UserAgent;
      my $req = new HTTP::Request('GET', 'https://www.example.com');
      my $res = $ua->request($req);
      print $res->code."\n";

      # PROXY SUPPORT
      $ENV{HTTPS_PROXY} = 'http://proxy_hostname_or_ip:port';

      # PROXY_BASIC_AUTH
      $ENV{HTTPS_PROXY_USERNAME} = 'username';
      $ENV{HTTPS_PROXY_PASSWORD} = 'password';  

      # DEBUGGING SWITCH / LOW LEVEL SSL DIAGNOSTICS
      $ENV{HTTPS_DEBUG} = 1;

      # DEFAULT SSL VERSION
      $ENV{HTTPS_VERSION} = '3';

      # CLIENT CERT SUPPORT
      $ENV{HTTPS_CERT_FILE} = 'certs/notacacert.pem';
      $ENV{HTTPS_KEY_FILE}  = 'certs/notacakeynopass.pem';

      # CA CERT PEER VERIFICATION
      $ENV{HTTPS_CA_FILE}   = 'certs/ca-bundle.crt';
      $ENV{HTTPS_CA_DIR}    = 'certs/';

      # CLIENT PKCS12 CERT SUPPORT
      $ENV{HTTPS_PKCS12_FILE}     = 'certs/pkcs12.pkcs12';
      $ENV{HTTPS_PKCS12_PASSWORD} = 'PKCS12_PASSWORD';

DESCRIPTION
    This perl module provides support for the https protocol under LWP, so
    that a LWP::UserAgent can make https GET & HEAD & POST requests. Please
    see perldoc LWP for more information on POST requests.

    The Crypt::SSLeay package contains Net::SSL, which is automatically
    loaded by LWP::Protocol::https on https requests, and provides the
    necessary SSL glue for that module to work via these deprecated modules:

       Crypt::SSLeay::CTX
       Crypt::SSLeay::Conn
       Crypt::SSLeay::X509

    Work on Crypt::SSLeay has been continued only to provide https support
    for the LWP - libwww perl libraries. If you want access to the OpenSSL
    API via perl, check out Sampo's Net::SSLeay.

INSTALL
  OpenSSL
    You must have OpenSSL or SSLeay installed before compiling this module.
    You can get the latest OpenSSL package from:

      http://www.openssl.org

    When installing openssl make sure your config looks like:

      > ./config --openssldir=/usr/local/openssl
     or
      > ./config --openssldir=/usr/local/ssl

    If you are planning on upgrading the default OpenSSL libraries on a
    system like RedHat, not that I would recommend this, then you might try
    something like:

      > ./config --openssldir=/usr --shared

    The --shared option to config will set up building the .so shared
    libraries which is important for such systems.

     then
      > make
      > make test
      > make install

    This way Crypt::SSLeay will pick up the includes and libraries
    automatically. If your includes end up going into a separate directory
    like /usr/local/include, then you may need to symlink
    /usr/local/openssl/include to /usr/local/include

  Crypt::SSLeay
    The latest Crypt::SSLeay can be found at your nearest CPAN, and also:

      http://www.perl.com/CPAN-local/modules/by-module/Crypt/

    Once you have downloaded it, Crypt::SSLeay installs easily using the
    make or nmake commands as shown below.

      > perl Makefile.PL
      > make
      > make test
      > make install

      * use nmake for win32

      !!! NOTE for Win32 users, few people seem to be able to build
      W  Crypt::SSLeay successfully on that platform.  You don't need
      I  to because ActiveState has already compiled it for you,
      N  and is available for their perl builds 618 & 522 as a ppm
      3  install.  It may also be available for their latest build.
      2  For problems with this, please contact ActiveState.
      !!!          Please see http://www.activestate.com/

PROXY SUPPORT
    LWP::UserAgent and Crypt::SSLeay have their own versions of proxy
    support. Please read these sections to see which one may be right for
    you.

  LWP::UserAgent Proxy Support
    LWP::UserAgent has its own methods of proxying which may work for you
    and is likely incompatible with Crypt::SSLeay proxy support. To use
    LWP::UserAgent proxy support, try something like:

      my $ua = new LWP::UserAgent;
      $ua->proxy([qw( https http )], "$proxy_ip:$proxy_port");

    At the time of this writing, libwww v5.6 seems to proxy https requests
    fine with an Apache mod_proxy server. It sends a line like:

      GET https://www.example.com HTTP/1.1

    to the proxy server, which is not the CONNECT request that some proxies
    would expect, so this may not work with other proxy servers than
    mod_proxy. The CONNECT method is used by Crypt::SSLeay's internal proxy
    support.

  Crypt::SSLeay Proxy Support
    For native Crypt::SSLeay proxy support of https requests, you need to
    set an environment variable HTTPS_PROXY to your proxy server & port, as
    in:

      # PROXY SUPPORT
      $ENV{HTTPS_PROXY} = 'http://proxy_hostname_or_ip:port';
      $ENV{HTTPS_PROXY} = '127.0.0.1:8080';

    Use of the HTTPS_PROXY environment variable in this way is similar to
    LWP::UserAgent->env_proxy() usage, but calling that method will likely
    override or break the Crypt::SSLeay support, so do not mix the two.

    Basic auth credentials to the proxy server can be provided this way:

      # PROXY_BASIC_AUTH
      $ENV{HTTPS_PROXY_USERNAME} = 'username';
      $ENV{HTTPS_PROXY_PASSWORD} = 'password';  

    For an example of LWP scripting with Crypt::SSLeay native proxy support,
    please see the source of the ./lwp-ssl-test script in the Crypt::SSLeay
    distribution.

CLIENT CERTIFICATE SUPPORT
    Certificate support is new provided by patches from Tobias Manthey. Is
    ALPHA as of .25, but looking pretty stable as of .29.

    PEM encoded certificate and private key files may be used like this:

      $ENV{HTTPS_CERT_FILE} = 'certs/notacacert.pem';
      $ENV{HTTPS_KEY_FILE}  = 'certs/notacakeynopass.pem';

    You may test your files with the ./net_ssl_test program by issuing a
    command like:

      ./net_ssl_test -cert=certs/notacacert.pem -key=certs/notacakeynopass.pem -d GET $HOST_NAME

    Additionally, if you would like to tell the client where the CA file is,
    you may set these. These *CA* configs are ALPHA as of version .29.

      $ENV{HTTPS_CA_FILE} = "some_file";
      $ENV{HTTPS_CA_DIR}  = "some_dir";

    There is no sample CA cert file at this time for testing, but you may
    configure ./net_ssl_test to use your CA cert with the -CAfile option.

  Creating a Test Certificate
    To create simple test certificates with openssl, you may:

         /usr/local/openssl/bin/openssl req -config /usr/local/openssl/openssl.cnf -new -days 365 -newkey rsa:1024 -x509 -keyout notacakey.pem -out notacacert.pem 

    To remove the pass phrase from the key file, execute this:
    /usr/local/openssl/bin/openssl rsa -in notacakey.pem -out
    notacakeynopass.pem

  PKCS12
    New as of version .45 is PKCS12 certificate support thanks to Daisuke
    Kuroda The directives for enabling use of these certificates is:

      $ENV{HTTPS_PKCS12_FILE}     = 'certs/pkcs12.pkcs12';
      $ENV{HTTPS_PKCS12_PASSWORD} = 'PKCS12_PASSWORD';

    Use of this type of certificate will take precedence over previous
    certificate settings described.

SSL VERSIONS
    Crypt::SSLeay tries very hard to connect to ANY SSL web server trying to
    accomodate servers that are buggy, old or simply not standards
    compliant. To this effect, this module will try SSL connections in this
    order:

      SSL v23  - should allow v2 & v3 servers to pick their best type
      SSL v3   - best connection type
      SSL v2   - old connection type

    Unfortunately, some servers seem not to handle a reconnect to SSL v3
    after a failed connect of SSL v23 is tried, so you may set before using
    LWP or Net::SSL:

      $ENV{HTTPS_VERSION} = 3;

    so that a SSL v3 connection is tried first. At this time only a SSL v2
    connection will be tried after this, as the connection attempt order
    remains unchanged by this setting.

COMPATIBILITY
    This module has been compiled on the following platforms:

     PLATFORM       CPU     SSL             PERL     VER    DATE            WHO
     --------       ---     ---             ----     ---    ----            ---
     Linux 2.4.7    x86     OpenSSL 0.9.7   5.00800  .51    2003-06-10      Joshua Chamas
     Linux 2.4.7    x86     OpenSSL 0.9.6g  5.00800  .49    2003-01-29      Joshua Chamas
     Win2000 SP2    x86     OpenSSL 0.9.7   5.00601  .49    2003-01-29      Joshua Chamas
     WinNT SP6      x86     OpenSSL 0.9.6a  5.00601  .45    2002-08-01      Joshua Chamas
     Linux 2.4.7    x86     OpenSSL 0.9.6d  5.00800  .45    2002-08-01      Joshua Chamas
     Linux 2.4.7    x86     OpenSSL 0.9.6   5.00601  .39    2002-06-23      Joshua Chamas
     Solaris 2.8    Sparc   ?               5.00503  .37    2002-05-31      Christopher Biow
     OpenBSD 2.8    Sparc   ?               5.00600  .25    2001-04-11      Tim Ayers
     Linux 2.2.14   x86     OpenSSL 0.9.6   5.00503  .25    2001-04-10      Joshua Chamas
     WinNT SP6      x86     OpenSSL 0.9.4   5.00404  .25    2001-04-10      Joshua Chamas
     Solaris 2.7    Sparc   OpenSSL 0.9.6   5.00503  .22    2001-03-01      Dave Paris
     AIX 4.3.2      RS/6000 OpenSSL 0.9.6   5.6.0    .19    2001-01-08      Peter Heimann
     Solaris 2.6    x86     OpenSSL 0.9.5a  5.00501  .17    2000-09-04      Joshua Chamas
     Linux 2.2.12   x86     OpenSSL 0.9.5a  5.00503  .16    2000-07-13      David Harris
     FreeBSD 3.2    ?x86    OpenSSL 0.9.2b  5.00503  ?      1999-09-29      Rip Toren
     Solaris 2.6    ?Sparc  OpenSSL 0.9.4   5.00404  ?      1999-08-24      Patrick Killelea
     FreeBSD 2.2.5  x86     OpenSSL 0.9.3   5.00404  ?      1999-08-19      Andy Lee
     Solaris 2.5.1  USparc  OpenSSL 0.9.4   5.00503  ?      1999-08-18      Marek Rouchal
     Solaris 2.6    x86     SSLeay 0.8.0    5.00501  ?      1999-08-12      Joshua Chamas
     Linux 2.2.10   x86     OpenSSL 0.9.4   5.00503  ?      1999-08-11      John Barrett
     WinNT SP4      x86     SSLeay 0.9.2    5.00404  ?      1999-08-10      Joshua Chamas

BUILD NOTES
  Win32, WinNT, Win2000, can't build
    If you cannot get it to build on your windows box, try ActiveState perl,
    at least their builds 522 & 618 are known to have a ppm install of
    Crypt::SSLeay available. Please see http://www.activestate.com for more
    info.

  AIX 4.3.2 - Symbol Error: __umoddi3 : referenced symbol not found
    The __umoddi3 problem applies here as well when compiling with gcc.

    Alternative solution: In Makefile.PL, prepend "-L"/usr/local/<path to
    your gcc lib>/<version> to the $LIBS value. Add after line 82:

     $LIBS = '-L' . dirname(`gcc -print-libgcc-file-name`) . ' ' . $LIBS;

  Solaris x86 - Symbol Error: __umoddi3 : referenced symbol not found
     Problem:

    On Solaris x86, the default PERL configuration, and preferred, is to use
    the ld linker that comes with the OS, not gcc. Unfortunately during the
    OpenSSL build process, gcc generates in libcrypto.a, from bn_word.c, the
    undefined symbol __umoddi3, which is supposed to be later resolved by
    gcc from libgcc.a

    The system ld linker does not know about libgcc.a by default, so when
    building Crypt::SSLeay, there is a linker error for __umoddi3

     Solution:

    The fix for this symlink your libgcc.a to some standard directory like
    /usr/local/lib, so that the system linker, ld, can find it when building
    Crypt::SSLeay.

  FreeBSD 2.x.x / Solaris - ... des.h:96 #error _ is defined ...
    If you encounter this error: "...des.h:96: #error _ is defined, but some
    strange definition the DES library cannot handle that...," then you need
    to edit the des.h file and comment out the "#error" line.

    Its looks like this error might be common to other operating systems,
    and that occurs with OpenSSL 0.9.3. Upgrades to 0.9.4 seem to fix this
    problem.

  SunOS 4.1.4, Perl 5.004_04 - ld.so: Undefined symbol: _CRYPT_mem_ctrl
    Problems: (initial build was fine, but execution of Perl scripts had
    problems)

    Got a message "ld.so: Undefined symbol: _CRYPT_mem_ctrl" solution: In
    the Makefile, comment out the line with "-fpic" (also try changing to
    "-fPIC", and this works also, not sure if one is preferred).

NOTES
    Many thanks to Gisle Aas for the original writing of this module and
    many others including libwww for perl. The web will never be the same :)

    Ben Laurie deserves kudos for his excellent patches for better error
    handling, SSL information inspection, and random seeding.

    Thanks to Dongqiang Bai for host name resolution fix when using a proxy.

    Thanks to Stuart Horner of Core Communications, Inc. who found the need
    for building --shared OpenSSL libraries.

    Thanks to Pavel Hlavnicka for a patch for freeing memory when using a
    pkcs12 file, and for inspiring more robust read() behavior.

    James Woodyatt is a champ for finding a ridiculous memory leak that has
    been the bane of many a Crypt::SSLeay user.

    Thanks to Bryan Hart for his patch adding proxy support, and thanks to
    Tobias Manthey for submitting another approach.

    Thanks to Alex Rhomberg for Alpha linux ccc patch.

    Thanks to Tobias Manthey for his patches for client certificate support.

    Thanks to Gamid Isayev for CA cert support and insight into error
    messaging.

    Thanks to Jeff Long for working through a tricky CA cert SSLClientVerify
    issue.

    Thanks to Chip Turner for patch to build under perl 5.8.0

    Thanks to Joshua Chamas for the time he spent maintaining the
    module.

SUPPORT
    For use of Crypt::SSLeay & Net::SSL with perl's LWP, please send email
    to libwww@perl.org .

    For OpenSSL or general SSL support please email the openssl user mailing
    list at openssl-users@openssl.org . This includes issues associated with
    building and installing OpenSSL on one's system.

    Please report all bugs at
    L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-SSLeay>.

    This module was originally written by Gisle Aas, and was subsequently
    maintained by Joshua Chamas.

    This module is currently maintained by David Landgren.

COPYRIGHT
     Copyright (c) 2006 David Landgren.
     Copyright (c) 1999-2003 Joshua Chamas.
     Copyright (c) 1998 Gisle Aas.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



CHANGES
0.53 

- 0.52_02 deemed stable

0.52_02 2006-12-20 19:29:01 UTC

- improved VMS support (CPAN bug #19829).
- add a test to see if cert file is readable in
  Net::SSL::configure_certs (CPAN bug #8498) and Net::SSL version
  to 2.78.
- known working platforms list removed from documentation. Too old,
  and CPAN Testers has the up-to-date information.
- minor documentation improvements.

0.52_01 2006-12-17

- add call to SSL_library_init() in new()
- maintenance taken over by brian d foy and David Landgren.

$MODULE=Crypt::SSLeay; $VERSION = .51; $DATE="2003-06-10";

- fixed build problem for OpenSSL 0.9.6 and some builds
  of perl 5.8.x which resulted in make error:

   /usr/include/openssl/des.h:193: parse error before '&' token"

  Thanks to Rob Brown for submitting a similar patch to cover this problem

- bug fix from Dongqiang Bai when server using proxy cannot
  resolve host name being connected to

+ Added documentation for updating system OpenSSL libraries 
  for systems such has RedHat that have shared libraries built
  Work sponsored by Stuart Horner of Core Communications, Inc.

$MODULE=Crypt::SSLeay; $VERSION = .49; $DATE="2003-01-30";

+ Documentation updates, including new support address
  for LWP issues, and $ENV{HTTPS_DEBUG} flag.

+ Added c:/openssl in default search path on win32 machines
  which is the recommended installation area in the openssl dist

+ Added patch from Pavel Hlavnicka for freeing memory leaks
  from SSL_CTX_use_pkcs12_file() whose functionality is triggered
  by the $ENV{HTTPS_PKCS12_*} settings

+ Set timeout to 15 seconds for ./net_ssl_test and lwp-ssl-test
  sample scripts for better testing of timeout behavior

+ Added alarm() during Net::SSL->read() to honor socket timeout setting
  for more robust applications.  read() will die_with_error() which
  in consistent with previous semantics used during SSL read() failure 
  Thanks to Pavel Hlavnicka for prompting this change.

+ Removed code that supported versions of SSLeay before version 0.8
  I believe SSLeay v.8 was released back in 1998

+ Added patch from Devin Heitmueller so that initial random seed
  would be taken from /dev/urandom if available via RAND_load_file API

$MODULE=Crypt::SSLeay; $VERSION = .45; $DATE="2002-08-01";

+ PKCS12 certificate support, patch submitted by Beni Takahashi,
  author of patch Daisuke Kuroda

+ Fixing compile warnings on Solaris 8/Sparc with Forte 7.0 about 
  implicit conversions and implicit declarations.  Thanks to 
  Marek Rouchal for bug report.

$MODULE=Crypt::SSLeay; $VERSION = .43; $DATE="2002-07-29";

- Removed unused dependency on URI::URL, thanks to Ric Steinberger
  for pointing out this problem under perl 5.8.0 

$MODULE=Crypt::SSLeay; $VERSION = .41; $DATE="2002-07-07";

- fixed t/net_ssl.t to work on Windows NT

$MODULE=Crypt::SSLeay; $VERSION = .40; $DATE="2002-07-03";

 + = improvement; - = bug fix

- fixed Makefile.PL use of dirname() which could error for perl 5.8.x
  Thanks to Chip Turner of RedHat for patch.

$MODULE=Crypt::SSLeay; $VERSION = .39; $DATE="2002-06-23";

- Fixed a runtime error with Net::SSL->proxy for running under
  perl warnings with no proxy defined, which t/net_ssl.t test case
  revealed.

+ Added t/net_ssl.t test for initializing a Net::SSL object

+ Added build platform success note for ( thanks Christopher! )

  Solaris 2.8    Sparc	?		5.00503	 .37	2002-05-31	Christopher Biow

+ Added build auto-detect for 0.9.6+ and only then use OPENSSL_free 
  instead of free() since older OpenSSL like 0.9.4 did not have it.

+ Added ./net_ssl_test -CAfile documentation, and root CA file from mod_ssl
  distribution at certs/ca-bundle.crt that can be used for general root CA
  peer certificate verification.

+ Added build notes for 

   SunOS 4.1.4, Perl 5.004_04 - ld.so: Undefined symbol: _CRYPT_mem_ctrl

  from Jeff Haferman.

+ When Net::SSL->connect() being called from LWP::UserAgent in proxy mode,
  will connect to the proxy passing the $ua->agent string as
    
   CONNECT $peer_addr:$peer_port HTTP/1.0
   User-Agent: $ua->agent

+ Integrated Richard Chen's patches for exposing the Net::SSL
  certificate dates via an API:

   Crypt::SSLeay::X509::not_before
   Crypt::SSLeay::X509::not_after

  which can be gotten to by a call like...

   Net::SSL->new(...)->get_peer_certificate->not_after;

  These methods return a normal timestamp like: 2002-05-22 11:15:17 GMT
  There is an example of its use in the ./net_ssl_test script.

  The Crypt::SSLeay::* modules will continue to remain undocumented
  because such functionality lies outside the scope of what this
  module is geared for.  Net::SSLeay provides a more general API for 
  OpenSSL functionality.

$MODULE=Crypt::SSLeay; $VERSION = .37; $DATE="2002-01-08";

 + = improvement; - = bug fix

- use OPENSSL_free() instead of free() to fix crashes with win32 perl 5.6.1
  Thanks to Doug MacEachern for patch.

+ added Makefile.PL Candidate() path for win32 builds of OpenSSL
  Thanks to David Morse for patch.

$MODULE=Crypt::SSLeay; $VERSION = .36; $DATE="2001-12-05";

- perl Makefile.PL C:/some_path should be picked up now ...
  was doing case insensitive drive letter check before

$MODULE=Crypt::SSLeay; $VERSION = .35; $DATE="2001-10-31";

+ Set local $SIG{PIPE} = \&die before $ssl->connect()
  to capture the "broken pipe" error associated with connecting
  to a computer that is not running a SSL web server,
  when Crypt::SSLeay is built with OpenSSL 0.9.6a.  This error did 
  not occur with OpenSSL 0.9.4 or OpenSSL 0.9.5a, but this fix should
  be compatible with those versions too.

$MODULE=Crypt::SSLeay; $VERSION = .33; $DATE="2001-10-31";

+ Documented differences / conflicts between LWP proxy support
  and Crypt::SSLeay which seems to be a source of confusion for users.

+ Added Net::SSL::get_peer_verify call so the warning header
  from LWP that says:

    Client-SSL-Warning: Peer certificate not verified

  can be suppressed when HTTPS_CA_FILE & HTTPS_CA_DIR environment

SSLeay.pm
package Crypt::SSLeay;

use Crypt::SSLeay::X509;

use strict;
use vars qw(@ISA $VERSION %CIPHERS);

require DynaLoader;

@ISA = qw(DynaLoader);
$VERSION = '0.53';

bootstrap Crypt::SSLeay $VERSION;

use vars qw(%CIPHERS);
%CIPHERS = (
   'NULL-MD5'     => "No encryption with a MD5 MAC",
   'RC4-MD5'      => "128 bit RC4 encryption with a MD5 MAC",
   'EXP-RC4-MD5'  => "40 bit RC4 encryption with a MD5 MAC",
   'RC2-CBC-MD5'  => "128 bit RC2 encryption with a MD5 MAC",
   'EXP-RC2-CBC-MD5' => "40 bit RC2 encryption with a MD5 MAC",
   'IDEA-CBC-MD5' => "128 bit IDEA encryption with a MD5 MAC",
   'DES-CBC-MD5'  => "56 bit DES encryption with a MD5 MAC",
   'DES-CBC-SHA'  => "56 bit DES encryption with a SHA MAC",
   'DES-CBC3-MD5' => "192 bit EDE3 DES encryption with a MD5 MAC",
   'DES-CBC3-SHA' => "192 bit EDE3 DES encryption with a SHA MAC",
   'DES-CFB-M1'   => "56 bit CFB64 DES encryption with a one byte MD5 MAC",
);


# A xsupp bug made this nessesary
sub Crypt::SSLeay::CTX::DESTROY  { shift->free; }
sub Crypt::SSLeay::Conn::DESTROY { shift->free; }
sub Crypt::SSLeay::X509::DESTROY { shift->free; }

1;

__END__

=head1 NAME

Crypt::SSLeay - OpenSSL glue that provides LWP https support

=head1 SYNOPSIS

  lwp-request https://www.example.com

  use LWP::UserAgent;
  my $ua = new LWP::UserAgent;
  my $req = new HTTP::Request('GET', 'https://www.example.com');
  my $res = $ua->request($req);
  print $res->code."\n";

  # proxy support
  $ENV{HTTPS_PROXY} = 'http://proxy_hostname_or_ip:port';

  # proxy_basic_auth
  $ENV{HTTPS_PROXY_USERNAME} = 'username';
  $ENV{HTTPS_PROXY_PASSWORD} = 'password';  

  # debugging (SSL diagnostics)
  $ENV{HTTPS_DEBUG} = 1;

  # default ssl version
  $ENV{HTTPS_VERSION} = '3';

  # client certificate support
  $ENV{HTTPS_CERT_FILE} = 'certs/notacacert.pem';
  $ENV{HTTPS_KEY_FILE}  = 'certs/notacakeynopass.pem';

  # CA cert peer verification
  $ENV{HTTPS_CA_FILE}   = 'certs/ca-bundle.crt';
  $ENV{HTTPS_CA_DIR}    = 'certs/';

  # Client PKCS12 cert support
  $ENV{HTTPS_PKCS12_FILE}     = 'certs/pkcs12.pkcs12';
  $ENV{HTTPS_PKCS12_PASSWORD} = 'PKCS12_PASSWORD';

=head1 DESCRIPTION

This document describes C<Crypt::SSLeay> version 0.53, released
2006-12-26.

This perl module provides support for the https protocol under LWP,
to allow an C<LWP::UserAgent> object to perform GET, HEAD and POST
requests. Please see LWP for more information on POST requests.

The C<Crypt::SSLeay> package provides C<Net::SSL>, which is loaded
by C<LWP::Protocol::https> for https requests and provides the
necessary SSL glue.

This distribution also makes following deprecated modules available:

  Crypt::SSLeay::CTX
  Crypt::SSLeay::Conn
  Crypt::SSLeay::X509

Work on Crypt::SSLeay has been continued only to
provide https support for the LWP (libwww-perl)
libraries. People wishing to access the OpenSSL API
directly from Perl are advised to look at the
C<Net::SSLeay> module.

  http://search.cpan.org/dist/Net_SSLeay.pm/

=head1 INSTALL

=head2 OpenSSL

You must have OpenSSL or SSLeay installed before compiling 
this module.  You can get the latest OpenSSL package from:

  http://www.openssl.org/

When installing openssl make sure your config looks like:

  > ./config --openssldir=/usr/local/openssl
 or
  > ./config --openssldir=/usr/local/ssl

If you are planning on upgrading the default OpenSSL libraries on 
a system like RedHat, not that I would recommend this, then 
you might try something like:

  > ./config --openssldir=/usr --shared

The --shared option to config will set up building the .so 
shared libraries which is important for such systems.

 then
  > make
  > make test
  > make install

This way Crypt::SSLeay will pick up the includes and 
libraries automatically.  If your includes end up
going into a separate directory like /usr/local/include,
then you may need to symlink /usr/local/openssl/include
to /usr/local/include

=head2 Crypt::SSLeay

The latest Crypt::SSLeay can be found at your nearest CPAN,
and also:

  http://search.cpan.org/dist/Crypt-SSLeay/

Once you have downloaded it, Crypt::SSLeay installs easily 
using the make or nmake commands as shown below.  

  > perl Makefile.PL
  > make
  > make test
  > make install

  * use nmake for win32

  !!! NOTE for Win32 users, few people seem to be able to build
  W  Crypt::SSLeay successfully on that platform.  You don't need
  I  to because ActiveState has already compiled it for you,
  N  and is available for their perl builds 618 & 522 as a ppm
  3  install.  It may also be available for their latest build.
  2  For problems with this, please contact ActiveState.
  !!!          Please see http://www.activestate.com/

=head1 PROXY SUPPORT

LWP::UserAgent and Crypt::SSLeay have their own versions of 
proxy support. Please read these sections to see which one
may be right for you.

=head2 LWP::UserAgent Proxy Support

LWP::UserAgent has its own methods of proxying which may work for
you and is likely to be incompatible with Crypt::SSLeay proxy support.
To use LWP::UserAgent proxy support, try something like:

  my $ua = new LWP::UserAgent;
  $ua->proxy([qw( https http )], "$proxy_ip:$proxy_port");

At the time of this writing, libwww v5.6 seems to proxy https 
requests fine with an Apache mod_proxy server.  It sends a line like:

  GET https://www.example.com HTTP/1.1

to the proxy server, which is not the CONNECT request that
some proxies would expect, so this may not work with other
proxy servers than mod_proxy.  The CONNECT method is used
by Crypt::SSLeay's internal proxy support.

=head2 Crypt::SSLeay Proxy Support

For native Crypt::SSLeay proxy support of https requests,
you need to set an environment variable HTTPS_PROXY to your 
proxy server and port, as in:

  # proxy support
  $ENV{HTTPS_PROXY} = 'http://proxy_hostname_or_ip:port';
  $ENV{HTTPS_PROXY} = '127.0.0.1:8080';

Use of the C<HTTPS_PROXY> environment variable in this way 
is similar to LWP::UserAgent->env_proxy() usage, but calling
that method will likely override or break the Crypt::SSLeay
support, so do not mix the two.

Basic auth credentials to the proxy server can be provided 
this way:

  # proxy_basic_auth
  $ENV{HTTPS_PROXY_USERNAME} = 'username';
  $ENV{HTTPS_PROXY_PASSWORD} = 'password';  

For an example of LWP scripting with Crypt::SSLeay native proxy
support, please look at the F<lwp-ssl-test> script in the 
Crypt::SSLeay distribution.

=head1 CLIENT CERTIFICATE SUPPORT

Client certificates are supported. PEM0encoded certificate and
private key files may be used like this:

  $ENV{HTTPS_CERT_FILE} = 'certs/notacacert.pem';
  $ENV{HTTPS_KEY_FILE}  = 'certs/notacakeynopass.pem';

You may test your files with the F<net_ssl_test> program,
bundled with the distribution, by issuing a command like:

  ./net_ssl_test -cert=certs/notacacert.pem \
	-key=certs/notacakeynopass.pem -d GET $HOST_NAME

Additionally, if you would like to tell the client where
the CA file is, you may set these.

  $ENV{HTTPS_CA_FILE} = "some_file";
  $ENV{HTTPS_CA_DIR}  = "some_dir";

There is no sample CA cert file at this time for testing,
but you may configure ./net_ssl_test to use your CA cert
with the -CAfile option. (TODO: then what is teh ./certs
directory in the distribution??)

=head2 Creating a test certificate

To create simple test certificates with OpenSSL, you may
run the following command:

  openssl req -config /usr/local/openssl/openssl.cnf \
    -new -days 365 -newkey rsa:1024 -x509 \
    -keyout notacakey.pem -out notacacert.pem 

To remove the pass phrase from the key file, run:

  openssl rsa -in notacakey.pem -out notacakeynopass.pem

=head2 PKCS12 support

The directives for enabling use of PKCS12 certificates is:

  $ENV{HTTPS_PKCS12_FILE}     = 'certs/pkcs12.pkcs12';
  $ENV{HTTPS_PKCS12_PASSWORD} = 'PKCS12_PASSWORD';

Use of this type of certificate takes precedence over previous
certificate settings described. (TODO: unclear? Meaning "the
presence of this type of certificate??)

=head1 SSL versions

Crypt::SSLeay tries very hard to connect to I<any> SSL web server
accomodating servers that are buggy, old or simply
not standards-compliant. To this effect, this module will
try SSL connections in this order:

  SSL v23 - should allow v2 and v3 servers to pick their best type
  SSL v3  - best connection type
  SSL v2  - old connection type

Unfortunately, some servers seem not to handle a reconnect
to SSL v3 after a failed connect of SSL v23 is tried,
so you may set before using LWP or Net::SSL:

  $ENV{HTTPS_VERSION} = 3;

so that a SSL v3 connection is tried first. At this time
only a SSL v2 connection will be tried after this, as the 
connection attempt order remains unchanged by this setting.

=head1 BUILD NOTES

=head2 Win32, WinNT, Win2000, can't build

If you cannot get it to build on your windows box, try 
ActiveState perl, at least their builds 522 & 618 are
known to have a ppm install of Crypt::SSLeay available.
Please see http://www.activestate.com for more info.

=head2 AIX 4.3.2 - Symbol Error: __umoddi3 : referenced symbol not found

The __umoddi3 problem applies here as well when compiling with gcc.

Alternative solution:
In Makefile.PL, prepend C<-L>/usr/local/<path to your gcc lib>/<version>
to the $LIBS value. Add after line 82:

 $LIBS = '-L' . dirname(`gcc -print-libgcc-file-name`) . ' ' . $LIBS;

=head2 Solaris x86 - Symbol Error: __umoddi3 : referenced symbol not found

 Problem:

On Solaris x86, the default PERL configuration, and preferred, is to use
the ld linker that comes with the OS, not gcc.  Unfortunately during the 
OpenSSL build process, gcc generates in libcrypto.a, from bn_word.c,
the undefined symbol __umoddi3, which is supposed to be later resolved
by gcc from libgcc.a

The system ld linker does not know about libgcc.a by default, so 
when building Crypt::SSLeay, there is a linker error for __umoddi3

 Solution:

The fix for this symlink your libgcc.a to some standard directory
like /usr/local/lib, so that the system linker, ld, can find
it when building Crypt::SSLeay.  

=head2 FreeBSD 2.x.x / Solaris - ... des.h:96 #error _ is defined ...

If you encounter this error: "...des.h:96: #error _ is
defined, but some strange definition the DES library cannot handle
that...," then you need to edit the des.h file and comment out the 
"#error" line.

Its looks like this error might be common to other operating
systems, and that occurs with OpenSSL 0.9.3.  Upgrades to
0.9.4 seem to fix this problem.

=head2 SunOS 4.1.4, Perl 5.004_04 - ld.so: Undefined symbol: _CRYPT_mem_ctrl

Problems: (initial build was fine, but execution of Perl scripts had problems)

Got a message "ld.so: Undefined symbol: _CRYPT_mem_ctrl"
solution:  In the Makefile, comment out the line with
"-fpic"  (also try changing to "-fPIC", and this works
also, not sure if one is preferred).

=head1 NOTES

Many thanks to Gisle Aas for the original writing of 
this module and many others including libwww for perl.  
The web will never be the same :)

Ben Laurie deserves kudos for his excellent patches
for better error handling, SSL information inspection,
and random seeding.

Thanks to Dongqiang Bai for host name resolution fix when
using a proxy.

Thanks to Stuart Horner of Core Communications, Inc. who found
the need for building --shared OpenSSL libraries.

Thanks to Pavel Hlavnicka for a patch for freeing memory when
using a pkcs12 file, and for inspiring more robust read() behavior.

James Woodyatt is a champ for finding a ridiculous memory
leak that has been the bane of many a Crypt::SSLeay user.

Thanks to Bryan Hart for his patch adding proxy support,
and thanks to Tobias Manthey for submitting another approach.

Thanks to Alex Rhomberg for Alpha linux ccc patch.

Thanks to Tobias Manthey for his patches for client 
certificate support.

Thanks to Daisuke Kuroda for adding PKCS12 certificate
support.

Thanks to Gamid Isayev for CA cert support and 
insights into error messaging.

Thanks to Jeff Long for working through a tricky CA
cert SSLClientVerify issue.

Thanks to Chip Turner for patch to build under perl 5.8.0.

Thanks to Joshua Chamas for the time he spent maintaining the
module.

=head1 SUPPORT

For use of Crypt::SSLeay & Net::SSL with perl's LWP, please
send email to C<libwww@perl.org>.

For OpenSSL or general SSL support please email the 
openssl user mailing list at C<openssl-users@openssl.org>.
This includes issues associated with building and installing
OpenSSL on one's system.

Please report all bugs at
L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-SSLeay>.

This module was originally written by Gisle Aas, and was subsequently
maintained by Joshua Chamas.

This module is currently maintained by David Landgren.

=head1 COPYRIGHT

 Copyright (c) 2006 David Landgren.
 Copyright (c) 1999-2003 Joshua Chamas.
 Copyright (c) 1998 Gisle Aas.

This program is free software; you can redistribute 
it and/or modify it under the same terms as Perl itself. 

=cut
